File: Week_06/714/best_time_to_buy_and_sell_stock_with_transaction_fee.go
package main

func max(x, y int) int {
	if x > y {
		return x
	}
	return y
}

func maxProfit(prices []int, fee int) int {
	hold, notHold := -prices[0]-fee, 0
	for i := 1; i < len(prices); i++ {
		notHold, hold = max(notHold, hold+prices[i]),max(hold, notHold-prices[i]-fee)
	}
	return notHold
}

File: Week_06/714/best_time_to_buy_and_sell_stock_with_transaction_fee.py
from typing import List


class Solution:
    def maxProfit(self, prices: List[int], fee: int) -> int:
        hold, not_hold = -prices[0] - fee, 0
        for p in prices[1:]:
            hold, not_hold = max(hold, not_hold - p - fee), max(not_hold, hold + p)
        return not_hold
